core/src/main/scala/org/apache/spark/scheduler/ResultTask.scala

```diff
@@ -66,7 +67,10 @@ private[spark] class ResultTask[T, U](
     _executorDeserializeTime = System.currentTimeMillis() - deserializeStartTime
 
     metrics = Some(context.taskMetrics)
-    func(context, rdd.iterator(partition, context))
+    val itr = rdd.iterator(partition, context)
+    val result = func(context, itr)
+    val computedFullPartition = itr.isEmpty || rdd.storageLevel != StorageLevel.NONE
```

Comment:
great.  would also like to see a test case that fails without this being addressed.  Sounds like you know how to workaround this, I was worried it might not be fixable.